Use formatting context root's writing mode and text orientation when synthesizing baselines.

sammygill opened the pull request:
#### c994ed0a39843aa33c3e125b9b3379ec90bc02f4
<pre>
Use formatting context root&apos;s writing mode and text orientation when synthesizing baselines.
<a href="https://bugs.webkit.org/show_bug.cgi?id=247815">https://bugs.webkit.org/show_bug.cgi?id=247815</a>
rdar://102245183

Reviewed by Alan Baradlay.

When synthesizing baselines for boxes we must take into consideration
the writing mode and text orientation of the formatting context root to
synthesize the correct type of baseline. The formatting context is also
used to determine which edges (content, border, or margin) of the box
to use.

The formatting context root&apos;s writing mode is used to determine which
type of baseline to synthesize. We could either synthesize the
alphabetic baseline or the central baseline depending on the scenario.
Previously, we were only synthesizing the alphabetic baseline. Now if
the writing mode is vertical with a text-orientation of &quot;mixed,&quot; or
&quot;upright,&quot; we will synthesize the central baseline.

Spec references:
<a href="https://drafts.csswg.org/css-align-3/#baseline-export">https://drafts.csswg.org/css-align-3/#baseline-export</a>
<a href="https://drafts.csswg.org/css-inline-3/#alignment-baseline-property">https://drafts.csswg.org/css-inline-3/#alignment-baseline-property</a>
<a href="https://drafts.csswg.org/css-inline-3/#dominant-baseline-property">https://drafts.csswg.org/css-inline-3/#dominant-baseline-property</a>

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/alignment/flex-align-baseline-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/alignment/flex-align-baseline-003-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/alignment/flex-align-baseline-006-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/alignment/flex-align-baseline-007-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/alignment/flex-align-baseline-line-clamp-002.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/alignment/flex-align-baseline-line-clamp-003.tentative-expected.txt:
* Source/WebCore/platform/graphics/FontBaseline.h:
* Source/WebCore/rendering/RenderBlock.cpp:
(WebCore::RenderBlock::inlineBlockBaseline const):
* Source/WebCore/rendering/RenderBlockFlow.cpp:
(WebCore::RenderBlockFlow::inlineBlockBaseline const):
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::synthesizedBaseline):
(WebCore::synthesizedBaselineFromBorderBox): Deleted.
* Source/WebCore/rendering/RenderBox.h:
* Source/WebCore/rendering/RenderFlexibleBox.cpp:
(WebCore::RenderFlexibleBox::baselinePosition const):
(WebCore::RenderFlexibleBox::firstLineBaseline const):
(WebCore::RenderFlexibleBox::lastLineBaseline const):
(WebCore::RenderFlexibleBox::marginBoxAscentForChild):
* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::RenderGrid::baselinePosition const):
(WebCore::RenderGrid::firstLineBaseline const):
(WebCore::RenderGrid::lastLineBaseline const):

Canonical link: <a href="https://commits.webkit.org/257239@main">https://commits.webkit.org/257239@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4b9b523347f87de54eabfa157eacfd633ce6c205

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/98177 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/7390 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/31322 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/107637 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/167899 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/102119 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/7884 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/36155 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/90765 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/104225 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/103829 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/5920 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/84773 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/33032 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/87802 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/89499 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/75969 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/1354 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/20931 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/1312 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/22434 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4987 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/6187 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/44891 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/2648 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/41856 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -5491,9 +5493,19 @@ LayoutBoxExtent RenderBox::scrollPaddingForViewportRect(const LayoutRect& viewpo
+
+    if (edge == Content)
+        boxSize -= direction == HorizontalLine ? box.verticalBorderAndPaddingExtent() : box.horizontalBorderAndPaddingExtent();
+    if (edge == Margin)
```
please use else if (as now it looks like edge is an option set)

sammygill replied:
Changed

alanbaradlay commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -5491,9 +5493,19 @@ LayoutBoxExtent RenderBox::scrollPaddingForViewportRect(const LayoutRect& viewpo
 {
-    return direction == HorizontalLine ? box.height() : box.width();
+    ASSERT(parentStyle);
+    LayoutUnit boxSize = direction == HorizontalLine ? box.height() : box.width();
```
please use auto boxSize

sammygill replied:
Changed

alanbaradlay commented on Source/WebCore/platform/graphics/FontBaseline.h:
```diff
@@ -28,7 +28,9 @@
-enum FontBaseline { AlphabeticBaseline, IdeographicBaseline };
+enum FontBaseline { AlphabeticBaseline, IdeographicBaseline, CentralBaseline };
+
+enum BaselineSynthesisEdge { Content, Border, Margin };
```
I assume these refer to the CSS box which case they should read { ContentBox, BorderBox, MarginBox }

sammygill replied:
Changed

alanbaradlay commented on Source/WebCore/rendering/RenderBlockFlow.cpp:
```diff
@@ -3049,7 +3049,7 @@ std::optional<LayoutUnit> RenderBlockFlow::inlineBlockBaseline(LineDirectionMode
     }
     // Note that here we only take the left and bottom into consideration. Our caller takes the right and top into consideration.
-    float boxHeight = synthesizedBaselineFromBorderBox(*this, lineDirection) + (lineDirection == HorizontalLine ? m_marginBox.bottom() : m_marginBox.left());
+    float boxHeight = synthesizedBaseline(*this, parentStyle(), lineDirection, Border) + (lineDirection == HorizontalLine ? m_marginBox.bottom() : m_marginBox.left());
```
any particular reason for this to be float?

sammygill replied:
Not as far as I can tell. Changed this to auto and changed `lastBaseline` below it to LayoutUnit

alanbaradlay commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -5491,9 +5493,19 @@ LayoutBoxExtent RenderBox::scrollPaddingForViewportRect(const LayoutRect& viewpo
 }
 
-LayoutUnit synthesizedBaselineFromBorderBox(const RenderBox& box, LineDirectionMode direction)
+LayoutUnit synthesizedBaseline(const RenderBox& box, const RenderStyle* parentStyle, LineDirectionMode direction, BaselineSynthesisEdge edge)
```
are we expecting a null_ptr parent style here? (judging by the assert below, the answer is no -and also during layout there should always be a parent()->style() unless we call this function on the ICB, which I think we should never....so shouldn't we just pass in a const ref instead?)

sammygill replied:
Should we move the null check to the caller instead or is there a better way to get this style information? Here my idea was to fallback to the old behavior in the strange scenario where parent style is null for some reason.

IF we move the null check to the caller we could probably still fall back to the old logic outside of this method and be able to pass in a const ref. Although, moving that fallback behavior to the caller would look a little strange and would need to be done in several places. Thoughts?

sammygill replied:
Changed

sammygill committed:
Use formatting context root's writing mode and text orientation when synthesizing baselines.
https://bugs.webkit.org/show_bug.cgi?id=247815
rdar://102245183

Reviewed by Alan Baradlay.

When synthesizing baselines for boxes we must take into consideration
the writing mode and text orientation of the formatting context root to
synthesize the correct type of baseline. The formatting context is also
used to determine which edges (content, border, or margin) of the box
to use.

The formatting context root's writing mode is used to determine which
type of baseline to synthesize. We could either synthesize the
alphabetic baseline or the central baseline depending on the scenario.
Previously, we were only synthesizing the alphabetic baseline. Now if
the writing mode is vertical with a text-orientation of "mixed," or
"upright," we will synthesize the central baseline.

Spec references:
https://drafts.csswg.org/css-align-3/#baseline-export
https://drafts.csswg.org/css-inline-3/#alignment-baseline-property
https://drafts.csswg.org/css-inline-3/#dominant-baseline-property

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/alignment/flex-align-baseline-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/alignment/flex-align-baseline-003-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/alignment/flex-align-baseline-006-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/alignment/flex-align-baseline-007-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/alignment/flex-align-baseline-line-clamp-002.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/alignment/flex-align-baseline-line-clamp-003.tentative-expected.txt:
* Source/WebCore/platform/graphics/FontBaseline.h:
* Source/WebCore/rendering/RenderBlock.cpp:
(WebCore::RenderBlock::inlineBlockBaseline const):
* Source/WebCore/rendering/RenderBlockFlow.cpp:
(WebCore::RenderBlockFlow::inlineBlockBaseline const):
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::synthesizedBaseline):
(WebCore::synthesizedBaselineFromBorderBox): Deleted.
* Source/WebCore/rendering/RenderBox.h:
* Source/WebCore/rendering/RenderFlexibleBox.cpp:
(WebCore::RenderFlexibleBox::baselinePosition const):
(WebCore::RenderFlexibleBox::firstLineBaseline const):
(WebCore::RenderFlexibleBox::lastLineBaseline const):
(WebCore::RenderFlexibleBox::marginBoxAscentForChild):
* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::RenderGrid::baselinePosition const):
(WebCore::RenderGrid::firstLineBaseline const):
(WebCore::RenderGrid::lastLineBaseline const):

Canonical link: https://commits.webkit.org/257239@main
